curvefs/src/client/s3/client_s3_cache_manager.cpp

```diff
@@ -0,0 +1,1092 @@
+    WriteLockGuard writeLockGuard(rwLockRead_);
+
+    dataRCacheMap_.emplace(chunkPos, iter);
+    return dataCache;
+}
+
+void ChunkCacheManager::ReleaseReadDataCache(uint64_t key) {
+    WriteLockGuard writeLockGuard(rwLockRead_);
+
+    auto iter = dataRCacheMap_.find(key);
+    if (iter == dataRCacheMap_.end()) {
+        return;
+    }
+
+    dataRCacheMap_.erase(iter);
+    return;
+}
+
+void ChunkCacheManager::ReleaseCache(S3ClientAdaptorImpl* s3ClientAdaptor) {
+    {
+        WriteLockGuard writeLockGuard(rwLockWrite_);
+        auto size = dataWCacheMap_.size();
+        dataWCacheMap_.clear();
+        if (size > 0) {
+            s3ClientAdaptor_->GetFsCacheManager()->DataCacheNumFetchSub(size);
+        }
+    }
+    WriteLockGuard writeLockGuard(rwLockRead_);
+    auto iter = dataRCacheMap_.begin();
+    for (; iter != dataRCacheMap_.end(); iter++) {
+        s3ClientAdaptor->GetFsCacheManager()->Delete(iter->second);
+    }
+    dataRCacheMap_.clear();
+}
+
+CURVEFS_ERROR ChunkCacheManager::Flush(Inode* inode, bool force) {
+    std::map<uint64_t, DataCachePtr> tmp;
+    CURVEFS_ERROR ret;
+    {
+        WriteLockGuard writeLockGuard(rwLockWrite_);
+        dataWCacheMap_.swap(tmp);
+    }
+    auto size = tmp.size();
+    auto iter = tmp.begin();
+    uint64_t i = 0;
+    for (; iter != tmp.end(); iter++) {
+        ret = iter->second->Flush(inode, force);
+        if (ret != CURVEFS_ERROR::OK) {
+            LOG(INFO) << "dataCache flush failed. ret:" << ret
+                      << ",index:" << index_
+                      << ",data chunkpos:" << iter->second->GetChunkPos();
+            if (dataWCacheMap_.empty()) {
+                WriteLockGuard writeLockGuard(rwLockWrite_);
+                dataWCacheMap_.swap(tmp);
+            } else {
+                // todo
+                s3ClientAdaptor_->GetFsCacheManager()
+                ->DataCacheNumFetchSub(size);
+            }
+            return ret;
+        }
+    }
+    s3ClientAdaptor_->GetFsCacheManager()->DataCacheNumFetchSub(size);
+    return CURVEFS_ERROR::OK;
+}
+
+void ChunkCacheManager::UpdateReadCache(uint64_t chunkPos, uint64_t len,
+                                        const char* dataBuf) {
+    {
+        WriteLockGuard writeLockGuard(rwLockRead_);
+
+        std::map<uint64_t, std::list<DataCachePtr>::iterator>::iterator iter =
+            dataRCacheMap_.begin();
+        for (; iter != dataRCacheMap_.end(); iter++) {
+            if (chunkPos + len < iter->first) {
+                break;
+            }
+            std::list<DataCachePtr>::iterator dcpIter = iter->second;
+            uint64_t dcChunkPos = (*dcpIter)->GetChunkPos();
+            uint64_t dcLen = (*dcpIter)->GetLen();
+            //  overlap, must delete
+            if ((chunkPos < dcChunkPos + dcLen) &&
+                (dcChunkPos < chunkPos + len)) {
+                s3ClientAdaptor_->GetFsCacheManager()->Delete(dcpIter);
+                dataRCacheMap_.erase(iter);
```

Comment:
I think it's a danguer operator for you traverse the map continue after erase the iterator.

Reply:
References and iterators to the erased elements are invalidated.

Reply:
fixed